client/peerset/src/lib.rs

```diff
@@ -615,165 +325,106 @@ impl Peerset {
-				// Decrease the node's reputation so that we don't try it again and again and again.
-				entry.add_reputation(DISCONNECT_REPUTATION_CHANGE);
-				trace!(target: "peerset", "Dropping {}: {:+} to {}",
-					peer_id, DISCONNECT_REPUTATION_CHANGE, entry.reputation());
-				entry.disconnect();
-			},
-			peersstate::Peer::NotConnected(_) | peersstate::Peer::Unknown(_) => {
-				error!(target: "peerset", "Received dropped() for non-connected node")
-			},
-		}
-
-		if let DropReason::Refused = reason {
-			self.on_remove_from_peers_set(set_id, peer_id);
-		}
-
-		self.alloc_slots(set_id);
+	pub fn dropped(&mut self, set_id: SetId, peer_id: PeerId, _reason: DropReason) {
+		self.protocol_handles[set_id.0].dropped(peer_id);
 	}
 
 	/// Reports an adjustment to the reputation of the given peer.
 	pub fn report_peer(&mut self, peer_id: PeerId, score_diff: ReputationChange) {
 		// We don't immediately perform the adjustments in order to have state consistency. We
 		// don't want the reporting here to take priority over messages sent using the
 		// `PeersetHandle`.
-		let _ = self.tx.unbounded_send(Action::ReportPeer(peer_id, score_diff));
+		let _ = self.to_self.unbounded_send(Action::ReportPeer(peer_id, score_diff));
 	}
 
 	/// Produces a JSON object containing the state of the peerset manager, for debugging purposes.
 	pub fn debug_info(&mut self) -> serde_json::Value {
-		self.update_time();
-
-		json!({
-			"sets": (0..self.data.num_sets()).map(|set_index| {
-				json!({
-					"nodes": self.data.peers().cloned().collect::<Vec<_>>().into_iter().filter_map(|peer_id| {
-						let state = match self.data.peer(set_index, &peer_id) {
-							peersstate::Peer::Connected(entry) => json!({
-								"connected": true,
-								"reputation": entry.reputation()
-							}),
-							peersstate::Peer::NotConnected(entry) => json!({
-								"connected": false,
-								"reputation": entry.reputation()
-							}),
-							peersstate::Peer::Unknown(_) => return None,
-						};
-
-						Some((peer_id.to_base58(), state))
-					}).collect::<HashMap<_, _>>(),
-					"reserved_nodes": self.reserved_nodes[set_index].0.iter().map(|peer_id| {
-						peer_id.to_base58()
-					}).collect::<HashSet<_>>(),
-					"reserved_only": self.reserved_nodes[set_index].1,
-				})
-			}).collect::<Vec<_>>(),
-			"message_queue": self.message_queue.len(),
-		})
+		// TODO: check what info we can include here.
+		json!("unimplemented")
 	}
 
 	/// Returns the number of peers that we have discovered.
 	pub fn num_discovered_peers(&self) -> usize {
-		self.data.peers().len()
+		self.peer_store_handle.num_known_peers()
 	}
 }
 
 impl Stream for Peerset {
 	type Item = Message;
 
 	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
-		loop {
-			if let Some(message) = self.message_queue.pop_front() {
-				return Poll::Ready(Some(message))
+		if let Poll::Ready(msg) = self.from_controllers.poll_next_unpin(cx) {
+			if let Some(msg) = msg {
```

Comment:
This could possibly be converted to use `map_or()` and the print could be removed?

Reply:
I'm going to keep it as it is for consistency with other prints in the function, if you are not ultimately against it.